### PR TITLE
Add initial support for themes to control the editor

### LINKED ISSUE #22655: Block Editor: Allow a theme to override global defaults for editor features
## Description

Related to #22622.

Follow up for #22291 where initial handling for block editor features were introduced.

Usage:

```js
// edit.js

const Edit = ( props ) => {
    const isDisabled = ! useEditorFeature( 'typography.dropCap' );
    // ...
};
```

At the moment this feature can be only controlled through the defaults set for the Gutenberg plugin in `lib/exeperimental-default-theme.json` file in the `features` section of the `global` settings. On the block level, it can be controlled through `supports`.

The idea is to take it to the next level – take & merge data from the `global` subtree of the `experimental-theme.json` with the existing data in `lib/exeperimental-default-theme.json`.

The full proposal from @nosolosw in https://github.com/WordPress/gutenberg/pull/22291#pullrequestreview-418481941.

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/22655
Follows-up https://github.com/WordPress/gutenberg/pull/22291
Related https://github.com/WordPress/gutenberg/issues/20588

This PR adds support for themes to control the editor via the `global` context.

It also merges the content of `editor-features.php` within `global-styles.php`, improving the performance as a result of having only one filesystem access to the same file by request.

## How to test

- Load this PR.
- Load edit-site and verify that the value `settings.__experimentalFeatures.typography.dropCap` at the `block-editor` store is `false`.
- Open the file `lib/experimental-theme.json` and set the dropcap value to `true` => verify the value is `true` in the store.
- Install and enable a theme with support for global styles (for example, the [demo theme](https://github.com/nosolosw/global-styles-theme/)). Set dropcap to `false` in the theme's `experimental-theme.json` file => verify the value is false in the store.

### CODE REVIEW

File: lib/global-styles.php [hunk 1]
```diff
@@ -535,40 +535,49 @@ function gutenberg_experimental_global_styles_normalize_schema( $tree ) {
...
- * core's, theme's, and user's origins.
- *
- * @return string
- */
-function gutenberg_experimental_global_styles_get_stylesheet() {
-	$gs_merged = array();
-	$gs_core   = gutenberg_experimental_global_styles_get_core();
-	$gs_theme  = gutenberg_experimental_global_styles_get_theme();
-	$gs_user   = gutenberg_experimental_global_styles_get_user();
+function gutenberg_experimental_global_styles_get_merged_origins() {
```
Comment (youknowriad) [inline lib/global-styles.php:538]: Can we add a PHP Doc to explain what this function is doing.

File: lib/global-styles.php [hunk 2]
```diff
@@ -535,40 +535,49 @@ function gutenberg_experimental_global_styles_normalize_schema( $tree ) {
...
+ */
+function gutenberg_experimental_global_styles_get_editor_features( $config ) {
+	if (
+		empty( $config['global']['features'] ) ||
+		! is_array( $config['global']['features'] )
+	) {
+		return array();
+	}
+
+	return $config['global']['features'];
```
Comment (youknowriad) [inline lib/global-styles.php:578]: I don't really think we should pass just the "global" features to the editor, we should pass all the features config (per block as well) and handle it in useFeature properly.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

File: lib/global-styles.php (hunk 1 above)
Comment (nosolosw) [reply lib/global-styles.php:544]: :+1: https://github.com/WordPress/gutenberg/pull/24275/commits/9aafb6e0f99ad1968e4740ce56326d4c2f4914cf

File: lib/global-styles.php (hunk 2 above)
Comment (nosolosw) [reply lib/global-styles.php:584]: Yup, that's my next step.

**Comment by AdsonCicilioti** [comment]
Wow!! What a coincidence!

I was looking for a solution to pass the styles of the theme to the Editor in the best possible way. The strategy I have pursued is very rude, as it hurts an override in several classes of the native blocks, so that they come in harmony with the theme of the site and with the blocks themselves.

This is bad because many updates come out and I can't keep up with the changes very well. In addition Editor's UI element as the floating controls are affected by the theme style, leaving them buggy.

Congratulations @nosolosw  for the initiative! If we can soon temify the styles of the blocks to conform to the themes and deliver a more faithful experience to the published content, it will be incredible!

Thank you very much in advance!

**Comment by AdsonCicilioti** [comment]
Ah! One more thing:

Currently, without this implementation, what would be the least rude strategy for overwriting block styles to conform to the theme's styles?

**Comment by nosolosw** [comment]
> Currently, without this implementation, what would be the least rude strategy for overwriting block styles to conform to the theme's styles?

Hey @AdsonCicilioti We have a few options, depending on your needs. There's some [theme supports](https://developer.wordpress.org/block-editor/developers/themes/theme-support/), there's the ability to [register new styles](https://developer.wordpress.org/block-editor/tutorials/javascript/extending-the-block-editor/) per block, there's [block patterns](https://developer.wordpress.org/block-editor/developers/block-api/block-patterns/), etc.

From a theme perspective, CSS specificity wars is one of the issues we hope this approach would fix, but it's also one that's quite difficult to talk about without a specific use case. Do you have one in mind? Is there an issue you're facing that you'd like to report?

**Comment by AdsonCicilioti** [comment]
Hello @nosolosw .. Thanks for your reply ...
I am using the block editor of the standard WP installation. His the same editor from Gutenberg Plugin?
I want to replace some CSS properties. The Column Block, for example. The questions from the Breakpoists media are very opinionated and do not work in harmony with theme style.

I'm trying to be as deconstructive as possible. I'm enabled in the editor's style and and using the editor's blocks to compose the content of my pages and posts, when I need to give visual conformity / harmony between the block elements and the theme elements, I use the CSS of my theme, which is loaded after the Block Editor Style. That way I can overwrite what I need to give the harmony.

The things I'm using to harmonize with my theme are:
- Customized Color Palette - I have them in Json objects that I share both via PHP, CSS and JS.
- Font sizes - Same thing as the Color palettes
- Theme styles in the editor - Basically the BADY background color, font-family, Block width and some auxiliary CSS classes.

The big problem as I see it is part of my css that takes care of overwriting some styles, it’s getting too big, not to mention that I’m carrying the common style of the blocks to the frontend plus the theme style with various properties that overwrite the styles of blocks.

I'm about to copy the common styles for my CSS files, modify them as needed, and disable it in the frontend.

The problem is that over time these common editor styles will change.

I do not want to resort to the creation of blocks that are very similar to the Editor's and that serve me in their structure. I will create new blocks, yes, but different blocks. It is inconsistent to recreate what already exists.

I really intend to fully explore the editor and its blocks, but it would be amazing if there was a better way to overwrite the styles.

**Pull request merged**